mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala

```diff
@@ -49,8 +51,45 @@ object HoverProvider:
     val exprTp = typeFromPath(enclosing)
     val exprTpw = exprTp.widenTermRefExpr
 
-    if tp.isError || tpw == NoType || tpw.isError || path.isEmpty then
-      ju.Optional.empty()
+    /**
+     * Check if the given `sourcePos` is on the name of enclosing tree.
+     * ```
+     * // For example, if the postion is on `foo`, returns true
+     * def foo(x: Int) = { ... }
+     *      ^
+     *
+     * // On the other hand, it points to non-name position, return false.
+     * def foo(x: Int) = { ... }
+     *  ^
+     * ```
+     * @param path - path to the position given by `Interactive.pathTo`
+     */
+    def isHoveringOnName(path: List[Tree], sourcePos: SourcePosition): Boolean =
+      def contains(tree: Tree): Boolean = tree match
+        case tree: NameTree =>
+          source.fold(false) { s =>
+            SourceTree(tree, s).namePos.contains(sourcePos)
+          }
+        // TODO: check the positions for NamedArg and Import
+        case _: NamedArg => true
+        case _: Import => true
+        case app: (Apply | TypeApply) => contains(app.fun)
+        case _ => false
+      end contains
+
+      val enclosing = path
+        .dropWhile(t => !t.symbol.exists && !t.isInstanceOf[NamedArg])
+        .headOption
+        .getOrElse(EmptyTree)
+      contains(enclosing)
+    end isHoveringOnName
+
+    if tp.isError || tpw == NoType || tpw.isError || path.isEmpty ||
+      (pos.start == pos.end && !isHoveringOnName(
+        enclosing,
+        pos
+      )) // don't check isHoveringOnName for RangeHover
```

Comment:
Could you extract this part of the condition to `def`?

Reply:
Refactored a bit in https://github.com/scalameta/metals/pull/3792/commits/cf020525fb33f675ac665bd090553bdee7db99c4